**學號：413180207**　｜　作業：Exercise20260924　｜　課程：電腦程式設計

# 習題：數學物件、Python 型別與 AI 中的數值表示

在數學中，我們經常接觸下列概念：

- 自然數（Natural Numbers）
- 整數（Integers）
- 有理數（Rational Numbers）
- 實數（Real Numbers）
- 複數（Complex Numbers）
- 向量（Vectors）
- 矩陣（Matrices）
- 集合（Sets）
- 命題的真假（Truth Values）

本習題將探討：

```text
數學物件
    ↓
Python 型別
    ↓
CPython 實作
    ↓
AI 中的數值表示
```

## 第一部分：Python 如何表示數學物件？

請查閱 Python 文件或自行測試，完成下表。

| 數學概念 | Python 型別 | 範例 |
|----------|------------|------|
| 自然數 | ? | ? |
| 整數 | ? | ? |
| 有理數 | ? | ? |
| 實數 | ? | ? |
| 複數 | ? | ? |
| 向量 | ? | ? |
| 矩陣 | ? | ? |
| 集合 | ? | ? |
| 命題真假 | ? | ? |

並回答下列問題：

1. 哪些是 Python 的內建型別（Built-in Types）？
2. 哪些需要額外模組才能較完整地表示？
3. Python 的型別分類與數學中的分類是否完全一致？

### ✍️ 第一部分：我的作答

先用程式實際測試，再整理成表格。`type.__module__ == 'builtins'` 代表它是 Python 的**內建型別**；否則代表它來自標準函式庫模組或第三方套件。

In [1]:
from fractions import Fraction
import numpy as np

samples = {
    "自然數 (N)":     5,
    "整數 (Z)":       -3,
    "有理數 (Q)":     Fraction(3, 4),
    "實數 (R)":       3.14159,
    "複數 (C)":       2 + 3j,
    "向量 (list)":    [1, 2, 3],
    "向量 (NumPy)":   np.array([1.0, 2.0, 3.0]),
    "矩陣 (list)":    [[1, 2], [3, 4]],
    "矩陣 (NumPy)":   np.array([[1, 2], [3, 4]]),
    "集合":           {1, 2, 3},
    "命題真假":       True,
}

print(f"{'數學概念':<14}{'範例':<24}{'型別':<26}{'內建?'}")
print("-" * 70)
for name, v in samples.items():
    t = type(v)
    builtin = "是" if t.__module__ == "builtins" else "否 (" + t.__module__.split('.')[0] + ")"
    print(f"{name:<14}{' '.join(repr(v).split()):<24}{t.__name__:<26}{builtin}")

數學概念          範例                      型別                        內建?
----------------------------------------------------------------------
自然數 (N)       5                       int                       是
整數 (Z)        -3                      int                       是
有理數 (Q)       Fraction(3, 4)          Fraction                  否 (fractions)
實數 (R)        3.14159                 float                     是
複數 (C)        (2+3j)                  complex                   是
向量 (list)     [1, 2, 3]               list                      是
向量 (NumPy)    array([1., 2., 3.])     ndarray                   否 (numpy)
矩陣 (list)     [[1, 2], [3, 4]]        list                      是
矩陣 (NumPy)    array([[1, 2], [3, 4]]) ndarray                   否 (numpy)
集合            {1, 2, 3}               set                       是
命題真假          True                    bool                      是


#### 完成後的表格

| 數學概念 | Python 型別 | 範例 | 是否內建 |
|----------|------------|------|----------|
| 自然數 | `int`（非負的 `int`） | `5`、`0` | ✅ 內建 |
| 整數 | `int` | `-3` | ✅ 內建 |
| 有理數 | `fractions.Fraction` | `Fraction(3, 4)` | ❌ 標準函式庫（需 `import fractions`） |
| 實數 | `float`（近似；精確十進位可用 `decimal.Decimal`） | `3.14159` | ✅ 內建（`Decimal` 不是） |
| 複數 | `complex` | `2+3j` | ✅ 內建 |
| 向量 | `list` / `tuple`（基本）；`numpy.ndarray`（一維，較完整） | `[1, 2, 3]`、`np.array([1., 2., 3.])` | `list`/`tuple` 內建；`ndarray` 是第三方 |
| 矩陣 | 巢狀 `list`（基本）；`numpy.ndarray`（二維，較完整） | `[[1, 2], [3, 4]]`、`np.array([[1, 2], [3, 4]])` | 巢狀 `list` 內建；`ndarray` 是第三方 |
| 集合 | `set`（可變）／`frozenset`（不可變） | `{1, 2, 3}` | ✅ 內建 |
| 命題真假 | `bool` | `True`、`False` | ✅ 內建 |

#### 問題回答

**1. 哪些是 Python 的內建型別？**
`int`（自然數、整數）、`float`（實數的近似）、`complex`（複數）、`set`／`frozenset`（集合）、`bool`（命題真假），以及 `list`／`tuple`（可用來粗略表示向量、矩陣）。

**2. 哪些需要額外模組才能較完整地表示？**
- 有理數：需要標準函式庫的 `fractions.Fraction`（精確的分數運算）。
- 精確十進位小數：需要 `decimal.Decimal`。
- 向量、矩陣：需要第三方套件 **NumPy**（`ndarray`），才有真正的向量加法、內積、矩陣乘法（`@`）、轉置、線性代數等運算。

**3. Python 的型別分類與數學中的分類是否完全一致？** —— **不一致**，下方程式會示範差異。

In [2]:
# 差異一：list 不是「向量」——+ 是串接，不是向量加法
print([1, 2] + [3, 4])                      # list：串接
print(np.array([1, 2]) + np.array([3, 4]))  # ndarray：真正的向量加法

# 差異二：bool 是 int 的子類別，True + True == 2（邏輯值卻能做算術）
print(True + True, issubclass(bool, int))

# 差異三：float 並不是「實數」——只是二進位有限小數的一個子集
print((0.1).as_integer_ratio())             # 0.1 實際存的是 m / 2**k 這種分數
import math
print(math.sqrt(2) ** 2 == 2)               # 無理數只能近似 → False

# 差異四：自然數沒有獨立型別，int 也不限制正負
print(type(5) is type(-5))

# 差異五：有理數不是內建，Fraction 才是精確的
print(1/3 * 3 == 1, Fraction(1, 3) * 3 == 1)

[1, 2, 3, 4]
[4 6]
2 True
(3602879701896397, 36028797018963968)
False
True
True True


**差異整理**

| 面向 | 數學 | Python |
|------|------|--------|
| 自然數 ⊂ 整數 | 兩個不同的集合 | 同一個型別 `int`（「非負」只是值的性質，不是型別） |
| 實數 | 包含無理數、無限精度 | `float` 只是 64 位元的近似，只能精確表示形如 m/2^k 的數 |
| 向量／矩陣 | 有加法、數乘、內積等運算結構 | 內建 `list` 只是「容器」，運算意義不同（`+` 是串接） |
| 真假值 | 與整數無關的邏輯值 | `bool` 是 `int` 的子類別，`True == 1` |
| 有理數 | 獨立的數系 | 不是內建，要靠 `fractions` 模組 |

所以：**Python 的型別是依「實作與運算」設計的，不是依數學的集合包含關係設計的。**

## 第二部分：數學概念與 Python 型別

數學上有下列包含關係：

$$
\mathbb N
\subseteq
\mathbb Z
\subseteq
\mathbb Q
\subseteq
\mathbb R
\subseteq
\mathbb C
$$

請回答：

1. Python 是否為上述每一個數系都提供獨立型別？
2. 為什麼自然數與整數都使用 `int` 表示？
3. Python 為什麼提供 `complex` 型別？
4. Python 為什麼沒有內建的有理數型別？
5. 下列三個物件的型別有何不同？

```python
1
1.0
1+0j
```

### ✍️ 第二部分：我的作答

In [3]:
import numbers
from fractions import Fraction

# Python 的「數字塔」(PEP 3141)：Number > Complex > Real > Rational > Integral
for v in [1, 1.0, 1 + 0j, Fraction(1, 2)]:
    print(f"{v!r:<12} {type(v).__name__:<10}",
          "Integral" if isinstance(v, numbers.Integral) else "",
          "Rational" if isinstance(v, numbers.Rational) else "",
          "Real"     if isinstance(v, numbers.Real)     else "",
          "Complex"  if isinstance(v, numbers.Complex)  else "")

1            int        Integral Rational Real Complex
1.0          float        Real Complex
(1+0j)       complex       Complex
Fraction(1, 2) Fraction    Rational Real Complex


**1. Python 是否為每一個數系都提供獨立型別？** —— **否。**
內建只有 `int`、`float`、`complex` 三種數值型別；自然數與整數共用 `int`；有理數靠標準函式庫 `Fraction`；實數則用 `float` 近似。Python 另在 `numbers` 模組用**抽象基底類別**表達 `Integral ⊂ Rational ⊂ Real ⊂ Complex` 的包含關係（見上面輸出），但那是「介面」，不是各自獨立的資料型別。

**2. 為什麼自然數與整數都使用 `int`？**
- 自然數是整數的子集合，用同一種表示法即可涵蓋，不需要多一個型別。
- 「非負」是值的**限制條件**，不是儲存方式的差異；若另設型別，`3 - 5` 的結果（-2）就會跑出自然數，反而要處理型別轉換與錯誤。
- 語言設計追求簡潔：只需一種整數型別，加上任意精度，就不用區分 `unsigned int`、`long` 等（像 C 語言那樣）。

**3. Python 為什麼提供 `complex` 型別？**
- 複數在工程與科學計算（訊號處理、傅立葉轉換、電路交流分析、量子力學）非常常用，內建可寫 `3+4j`，語法方便。
- 複數運算需要成對的實部／虛部一起計算，做成內建型別效率高、寫法直覺；`abs()`、`+ - * /`、`**` 都直接支援，搭配 `cmath` 可算 `sqrt(-1)` 等。
- 複數系（ℂ）對代數運算是封閉的（例如負數開根號），這是實數系做不到的。

**4. Python 為什麼沒有內建的有理數型別？**
- 大多數程式用 `int` 與 `float` 就足夠，有理數需求少，不值得放進語言核心。
- `Fraction` 可以完全用 `int`（分子、分母）在純 Python 裡實作，放在標準函式庫 `fractions` 即可，保持核心精簡（"batteries included" 但核心小）。
- 分數運算需要一直約分（gcd），比 `int`／`float` 慢，不適合當作預設的數值型別。

In [4]:
# 5. 三個物件的型別比較
for x in (1, 1.0, 1 + 0j):
    print(f"{x!r:<8} type = {type(x).__name__:<8} 大小 = {__import__('sys').getsizeof(x)} bytes")

print("\n相等嗎？", 1 == 1.0 == 1 + 0j)
print("hash 相同嗎？", hash(1) == hash(1.0) == hash(1 + 0j))
print("同一個物件嗎？", id(1) == id(1.0))

# 混合運算時，結果型別會「往上」提升：int → float → complex
print(type(1 + 1.0).__name__, type(1.0 + 1j).__name__, type(1 + 1j).__name__)
print(1 / 2, 1 // 2, 1.0 // 2)

# 只有 complex 能對負數開平方根
import cmath, math
print(cmath.sqrt(-1))
try:
    math.sqrt(-1)
except ValueError as e:
    print("math.sqrt(-1) →", e)

1        type = int      大小 = 28 bytes
1.0      type = float    大小 = 24 bytes
(1+0j)   type = complex  大小 = 32 bytes

相等嗎？ True
hash 相同嗎？ True
同一個物件嗎？ False
float complex complex
0.5 0 0.0
1j
math.sqrt(-1) → math domain error


**5. `1`、`1.0`、`1+0j` 的型別有何不同？**

| 物件 | 型別 | 內部儲存 | 特性 |
|------|------|----------|------|
| `1` | `int` | 任意精度的整數（一串 30-bit digit） | 精確、無小數部分 |
| `1.0` | `float` | 1 個 IEEE 754 雙精度浮點數（64 bits） | 有小數部分，但有捨入誤差與範圍限制 |
| `1+0j` | `complex` | 2 個雙精度浮點數（實部、虛部） | 可表示虛數，運算結果可能變成複數 |

三者**數學上相等**（`==` 為 `True`，甚至 `hash` 也相同，所以在 `dict`／`set` 裡會被視為同一個鍵），但它們是**三個不同的物件**，型別、記憶體大小、可用的運算都不一樣。混合運算時，型別會依 `int → float → complex` 的順序自動提升（`1 + 1.0` 得到 `float`）。

## 第三部分：CPython 如何實作這些型別？

除了知道 Python 使用哪些型別表示數學物件之外，

我們也希望了解：

> Python 物件在 CPython 中是如何實作的？

請查閱資料，完成下表。

| 數學概念 | Python 型別 | CPython 實作 | 主要特點 |
|----------|------------|-------------|----------|
| 自然數 | `int` | ? | ? |
| 整數 | `int` | ? | ? |
| 實數 | `float` | ? | ? |
| 複數 | `complex` | ? | ? |
| 集合 | `set` | ? | ? |
| 命題真假 | `bool` | ? | ? |
| 串列 | `list` | ? | ? |

### ✍️ 第三部分：我的作答

（以下以 **64 位元 CPython 3.12** 為準；不同版本的欄位細節略有差異，但整體設計相同。）

| 數學概念 | Python 型別 | CPython 實作 | 主要特點 |
|----------|------------|-------------|----------|
| 自然數 | `int` | `PyLongObject`（C 結構 `struct _longobject`）：`PyObject_HEAD` ＋ 記錄符號與位數的欄位 ＋ 變長陣列 `digit ob_digit[]` | 與整數共用；用非負值即可；每個 digit 為 30 bits（存在 32-bit 整數中）；小整數 -5～256 為預先建立好的單例（快取） |
| 整數 | `int` | 同上：`PyLongObject` | 任意精度（Arbitrary Precision）、不可變（immutable）；位數隨數值大小增長，受限於記憶體 |
| 實數 | `float` | `PyFloatObject`：`PyObject_HEAD` ＋ 一個 C 的 `double ob_fval` | IEEE 754 binary64（1+11+52 bits）；固定 8 bytes 資料；有捨入誤差、有範圍上下限；不可變 |
| 複數 | `complex` | `PyComplexObject`：`PyObject_HEAD` ＋ `Py_complex cval`（內含 `double real; double imag;`） | 由兩個 `double` 組成；固定大小；不可變；誤差性質與 `float` 相同 |
| 集合 | `set` | `PySetObject`：以**雜湊表**實作，含 `fill`、`used`、`mask`、`setentry *table`、內建小表 `smalltable[8]`；每個 `setentry` 存 `key` 指標與 `hash` | 元素必須可雜湊（hashable）、無序、不重複；查詢／加入／刪除平均 O(1)；可變 |
| 命題真假 | `bool` | `PyBoolObject`（在 C 中就是 `PyLongObject`）；全域只有兩個單例物件 `Py_True`、`Py_False` | 是 `int` 的子類別（`True==1`、`False==0`）；不能再被繼承；永遠只有這兩個物件 |
| 串列 | `list` | `PyListObject`：`PyObject_VAR_HEAD` ＋ `PyObject **ob_item`（指標陣列）＋ `Py_ssize_t allocated`（已配置容量） | 動態陣列，存的是「物件的指標」，可放不同型別；可變；以「過量配置」讓 `append` 平均 O(1) |

## 第四部分：深入探討各種型別

請針對第三部分中的每一種型別回答：

1. 它是否為 Python 的內建型別？
2. 對應的 CPython 物件結構名稱為何？
3. 它如何儲存資料？
4. 是否有固定大小限制？
5. 若資料規模持續增加，如何擴充儲存空間？

### ✍️ 第四部分：我的作答

先用程式觀察各型別的實際大小與成長方式，再逐一回答。

In [5]:
import sys

print("=== int：位數隨數值增加 ===")
print(sys.int_info)   # bits_per_digit=30, sizeof_digit=4
for n in [0, 1, 2**30 - 1, 2**30, 2**60, 2**300, 10**1000]:
    label = f"10**1000" if n == 10**1000 else (f"2**{n.bit_length()-1}" if n > 1 and n & (n - 1) == 0 else str(n))
    print(f"{label:>12}  bit_length={n.bit_length():>5}  sizeof={sys.getsizeof(n):>5} bytes")

print("\n=== float / complex / bool：固定大小 ===")
print("float(1.0)      :", sys.getsizeof(1.0), "bytes")
print("float(1e300)    :", sys.getsizeof(1e300), "bytes   ← 再大也一樣大")
print("complex(1+2j)   :", sys.getsizeof(1 + 2j), "bytes  (兩個 double)")
print("bool(True)      :", sys.getsizeof(True), "bytes")

print("\n=== bool 是單例、是 int 的子類別 ===")
print(bool(1) is True, bool(0) is False, isinstance(True, int), True == 1)

print("\n=== 小整數快取 (-5 ~ 256) ===")
print(int("256") is int("256"), int("257") is int("257"))

=== int：位數隨數值增加 ===
sys.int_info(bits_per_digit=30, sizeof_digit=4, default_max_str_digits=4300, str_digits_check_threshold=640)
           0  bit_length=    0  sizeof=   28 bytes
           1  bit_length=    1  sizeof=   28 bytes
  1073741823  bit_length=   30  sizeof=   28 bytes
       2**30  bit_length=   31  sizeof=   32 bytes
       2**60  bit_length=   61  sizeof=   36 bytes
      2**300  bit_length=  301  sizeof=   68 bytes
    10**1000  bit_length= 3322  sizeof=  468 bytes

=== float / complex / bool：固定大小 ===
float(1.0)      : 24 bytes
float(1e300)    : 24 bytes   ← 再大也一樣大
complex(1+2j)   : 32 bytes  (兩個 double)
bool(True)      : 28 bytes

=== bool 是單例、是 int 的子類別 ===
True True True True

=== 小整數快取 (-5 ~ 256) ===
True False


In [6]:
# list 的成長（過量配置 over-allocation）
L = []
base = sys.getsizeof(L)
last = base
print("list 容量變化（每個指標 8 bytes）")
print(f"len= 0  size={base}B  capacity=0")
for i in range(1, 41):
    L.append(i)
    s = sys.getsizeof(L)
    if s != last:
        print(f"len={i:>2}  size={s}B  capacity={(s - base) // 8}")
        last = s

# set 的成長（雜湊表擴張、重新雜湊）
S = set()
last = sys.getsizeof(S)
print(f"\nset 大小變化\nlen= 0  size={last}B")
for i in range(1, 2001):
    S.add(i)
    s = sys.getsizeof(S)
    if s != last:
        print(f"len={i:>4}  size={s}B")
        last = s

list 容量變化（每個指標 8 bytes）
len= 0  size=56B  capacity=0
len= 1  size=88B  capacity=4
len= 5  size=120B  capacity=8
len= 9  size=184B  capacity=16
len=17  size=248B  capacity=24
len=25  size=312B  capacity=32
len=33  size=376B  capacity=40

set 大小變化
len= 0  size=216B
len=   5  size=728B
len=  19  size=2264B
len=  77  size=8408B
len= 307  size=32984B
len=1229  size=131288B


#### 1️⃣ `int`（自然數／整數）
1. **是否內建？** 是。
2. **CPython 結構：** `PyLongObject`（`struct _longobject`）。
3. **如何儲存資料？** 用「變長陣列」`ob_digit[]` 存放數值，每個 digit 使用 32-bit 空間中的 **30 bits**（也就是以 2³⁰ 為基底的數字系統）；另有欄位記錄正負號與 digit 個數（3.12 之前是 `ob_size`，3.12 起併入 `lv_tag`）。
4. **固定大小限制？** **沒有固定位元數限制**（程式輸出中 `sizeof` 隨數值增加）；唯一限制是可用記憶體（與計算時間）。
5. **資料變大時如何擴充？** 需要更多 digit 時就配置**更大的物件**（因 `int` 不可變，運算結果是新物件），digit 陣列長度隨之增加。

#### 2️⃣ `float`（實數）
1. **是否內建？** 是。
2. **CPython 結構：** `PyFloatObject`（`PyObject_HEAD` + `double ob_fval`）。
3. **如何儲存？** 直接以 C 的 `double`（IEEE 754 binary64）存：1 位元正負號、11 位元指數、52 位元尾數。
4. **固定大小限制？** **有**。固定 64 bits（程式輸出：任何 float 都是 24 bytes 的物件），數值範圍約 ±1.8×10³⁰⁸，有效精度約 15～17 位十進位數。
5. **如何擴充？** **不能擴充**。超出範圍會變成 `inf`（或丟出 `OverflowError`），精度也不會變多。

#### 3️⃣ `complex`（複數）
1. **是否內建？** 是。
2. **CPython 結構：** `PyComplexObject`（`PyObject_HEAD` + `Py_complex cval`，其中 `Py_complex` 為 `{double real; double imag;}`）。
3. **如何儲存？** 實部與虛部各用一個 `double`。
4. **固定大小限制？** 有，2 × 64 bits，限制同 `float`。
5. **如何擴充？** 不能擴充。

#### 4️⃣ `set`（集合）
1. **是否內建？** 是。
2. **CPython 結構：** `PySetObject`，內部是**雜湊表**（open addressing），每個槽位是 `setentry {PyObject *key; Py_hash_t hash;}`。
3. **如何儲存？** 依元素的 `hash` 值算出位置（`hash & mask`）；碰撞時以探測（probing）找下一個空位。元素只存「物件指標」與其雜湊值。空集合先使用內建的 `smalltable[8]`（8 個槽位）。
4. **固定大小限制？** 沒有固定上限（受記憶體限制）；但**元素必須可雜湊**（`list`、`set` 不能放進 `set`）。
5. **如何擴充？** 當表太滿（約使用超過 3/5，即 `fill*5 >= mask*3`）就配置一張**更大的表**（元素較少時約 4 倍，較多時約 2 倍，大小為 2 的次方）並將全部元素**重新雜湊**放入——這就是上面輸出中 `size` 呈階梯式跳升的原因。

#### 5️⃣ `bool`（命題真假）
1. **是否內建？** 是（`int` 的子類別）。
2. **CPython 結構：** `PyBoolObject`，實際上就是 `PyLongObject`；全域僅有 `Py_True`、`Py_False` 兩個單例。
3. **如何儲存？** 與 `int` 相同：`False` 是數值 0，`True` 是數值 1。
4. **固定大小限制？** 有：只有兩個值、兩個物件（`bool(x)` 永遠回傳這兩個物件之一，所以 `is True` 成立）。
5. **如何擴充？** 不需要也不能擴充，且 `bool` 不能被繼承。

#### 6️⃣ `list`（串列）
1. **是否內建？** 是。
2. **CPython 結構：** `PyListObject`（`PyObject_VAR_HEAD` + `PyObject **ob_item` + `Py_ssize_t allocated`）。
3. **如何儲存？** 一塊連續的**指標陣列**，每個元素是「指向物件的指標」（所以可混放不同型別，但也不如 NumPy 陣列緊密）。`ob_size` 是目前元素個數，`allocated` 是已配置容量。
4. **固定大小限制？** 沒有固定上限（受記憶體限制）。
5. **如何擴充？** `append` 容量不足時，用 `realloc` 配置更大的指標陣列，且**多配置一些空間**（約 1.125 倍再加一點常數：`new_allocated = (newsize + (newsize >> 3) + 6) & ~3`），使多次 `append` 的**平均成本為 O(1)**——這就是上面輸出中容量 0→4→8→16→24→32→40… 的原因。

## 第五部分：Python 整數與浮點數

執行下列程式：

```python
print(10**100)
print(10**1000)
```

回答：

1. 為什麼 Python 能表示如此大的整數？
2. Python 的 `int` 是否具有固定的位元數限制？
3. 什麼是 Arbitrary Precision（任意精度）？
4. CPython 使用什麼物件來實作 `int`？
5. Python 整數大小真正的限制是什麼？

### ✍️ 第五部分：我的作答（整數）

In [7]:
print(10**100)
print(10**1000)

10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

In [8]:
x = 10**1000
print("位數（十進位）  :", len(str(x)))
print("位元數 bit_length:", x.bit_length())
print("物件大小 bytes   :", sys.getsizeof(x))
print("digit 個數 ≈     :", (sys.getsizeof(x) - 24) // 4)   # 每個 digit 4 bytes（30 bits 有效）

位數（十進位）  : 1001
位元數 bit_length: 3322
物件大小 bytes   : 468
digit 個數 ≈     : 111


In [9]:
# 整數大小真正的限制：記憶體、時間，以及「整數 ↔ 字串」轉換的安全上限
big = 10**5000
print("bit_length OK :", big.bit_length())
if hasattr(sys, "get_int_max_str_digits"):        # Python 3.11+
    print("預設字串轉換位數上限:", sys.get_int_max_str_digits())
    try:
        str(big)
    except ValueError as e:
        print("str(10**5000) 失敗 →", str(e)[:60], "...")
    sys.set_int_max_str_digits(0)                 # 0 = 不限制
    print("解除限制後，位數 =", len(str(big)))
    sys.set_int_max_str_digits(4300)              # 還原預設

bit_length OK : 16610
預設字串轉換位數上限: 4300
str(10**5000) 失敗 → Exceeds the limit (4300 digits) for integer string conversio ...
解除限制後，位數 = 5001


**1. 為什麼 Python 能表示如此大的整數？**
因為 Python 的 `int` 不是對應 CPU 的 32／64 位元暫存器，而是一個**可變長度的物件**：需要多少位數，就用多少個 digit 來儲存（`10**1000` 約需 3322 bits，約 111 個 30-bit digit）。運算時由 CPython 用軟體逐個 digit 進位計算。

**2. Python 的 `int` 是否具有固定的位元數限制？** —— **沒有**。C 語言的 `int32`／`int64` 有固定位寬，超出會溢位；Python 的 `int` 不會溢位，`sys.getsizeof` 會隨數值變大而增加。

**3. 什麼是 Arbitrary Precision（任意精度）？**
數值的精度（可精確表示的位數）**不預設上限**，可依需要增加，整數運算結果永遠是**精確**的，不會捨入或溢位。（限制只在記憶體與運算時間。）

**4. CPython 使用什麼物件來實作 `int`？**
`PyLongObject`（C 結構 `struct _longobject`）——以 2³⁰ 為基底的 digit 陣列 ＋ 記錄符號與長度的欄位。

**5. Python 整數大小真正的限制是什麼？**
- **可用記憶體**（每 30 bits 約需 4 bytes）以及位數 / 長度欄位所能表示的上限（理論上也是有限的）。
- **計算時間**：位數越多，乘法、除法越慢（CPython 大數乘法使用 Karatsuba 演算法，約 O(n^1.58)）。
- **轉換限制**：從 Python 3.11 起，為了避免拒絕服務攻擊，`int` ↔ `str`（十進位）預設限制 4300 位（可用 `sys.set_int_max_str_digits` 調整），這不是儲存限制，而是轉換限制。

再執行：

```python
0.1 + 0.2
```

```python
0.1 + 0.2 == 0.3
```

```python
format(0.1, ".20f")
```

回答：

1. 為什麼結果與數學上的答案不同？
2. 浮點數是否能精確表示所有小數？
3. 什麼是浮點數的精度（precision）？
4. 什麼是浮點數的範圍（range）？
5. 精度與範圍有何不同？

### ✍️ 第五部分：我的作答（浮點數）

In [10]:
0.1 + 0.2

0.30000000000000004

In [11]:
0.1 + 0.2 == 0.3

False

In [12]:
format(0.1, ".20f")

'0.10000000000000000555'

In [13]:
from decimal import Decimal
import math

print("0.1 實際儲存的精確值 :", Decimal(0.1))
print("0.2 實際儲存的精確值 :", Decimal(0.2))
print("0.1+0.2 實際結果     :", Decimal(0.1 + 0.2))
print("0.3 實際儲存的精確值 :", Decimal(0.3))
print("\n二進位分數表示 (分子, 分母)：", (0.1).as_integer_ratio())
print("\n改用 Fraction / Decimal / isclose 可避免問題：")
print(Fraction(1, 10) + Fraction(2, 10) == Fraction(3, 10))
print(Decimal("0.1") + Decimal("0.2") == Decimal("0.3"))
print(math.isclose(0.1 + 0.2, 0.3))
total = 0.0
for _ in range(10):
    total += 0.1
print("\n逐次累加 10 次 0.1 =", total, "（誤差累積，不等於 1.0）")
print("math.fsum 精確加總    =", math.fsum([0.1] * 10))

0.1 實際儲存的精確值 : 0.1000000000000000055511151231257827021181583404541015625
0.2 實際儲存的精確值 : 0.200000000000000011102230246251565404236316680908203125
0.1+0.2 實際結果     : 0.3000000000000000444089209850062616169452667236328125
0.3 實際儲存的精確值 : 0.299999999999999988897769753748434595763683319091796875

二進位分數表示 (分子, 分母)： (3602879701896397, 36028797018963968)

改用 Fraction / Decimal / isclose 可避免問題：
True
True
True

逐次累加 10 次 0.1 = 0.9999999999999999 （誤差累積，不等於 1.0）
math.fsum 精確加總    = 1.0


In [14]:
# 精度 (precision) 與範圍 (range)
fi = sys.float_info
print("最大值 max      :", fi.max)
print("最小正規化數 min:", fi.min)
print("機器 epsilon    :", fi.epsilon, "（1 與下一個可表示的數之差）")
print("尾數位元 mant_dig:", fi.mant_dig, "（含隱藏位元）→ 約", fi.dig, "位可靠的十進位有效數字\n")

print("精度示範：大數附近間距變大")
print(1e16 + 1 == 1e16, 2.0**53 + 1 == 2.0**53)
print("\n範圍示範：")
print(1e308 * 10)                 # inf
try:
    2.0 ** 1024
except OverflowError as e:
    print("2.0**1024 →", e)
print(10**400 > 0, "← int 沒問題")
try:
    float(10**400)
except OverflowError as e:
    print("float(10**400) →", e)

最大值 max      : 1.7976931348623157e+308
最小正規化數 min: 2.2250738585072014e-308
機器 epsilon    : 2.220446049250313e-16 （1 與下一個可表示的數之差）
尾數位元 mant_dig: 53 （含隱藏位元）→ 約 15 位可靠的十進位有效數字

精度示範：大數附近間距變大
True True

範圍示範：
inf
2.0**1024 → (34, 'Numerical result out of range')
True ← int 沒問題
float(10**400) → int too large to convert to float


**1. 為什麼結果與數學上的答案不同？**
電腦用二進位儲存。`0.1`、`0.2` 用二進位表示是**無限循環小數**（像 1/3 在十進位是 0.333…），只能存成最接近的 64 位元浮點數：`0.1` 實際是 `0.1000000000000000055511151231257827…`，`0.2` 也略大於 0.2。兩者相加再四捨五入到最近的 double，得到 `0.30000000000000004`，而它與 `0.3` 所存的 double 不是同一個，所以 `0.1 + 0.2 == 0.3` 為 `False`。這**不是 Python 的錯**，所有使用 IEEE 754 的語言結果都一樣。

**2. 浮點數是否能精確表示所有小數？** —— **不能**。只有形如 m / 2ᵏ（分母為 2 的次方）的數才能精確表示，例如 `0.5`、`0.25`、`0.125`；而 `0.1`、`0.3` 這類數都只是近似值。

**3. 什麼是浮點數的精度（precision）？**
指能區分的**相對細緻程度**，也就是有效位數的多寡。`float`（binary64）有 53 位元有效數字（尾數 52 位＋隱藏位元），約等於 15～17 位十進位有效數字，機器 epsilon 約為 2.2×10⁻¹⁶。

**4. 什麼是浮點數的範圍（range）？**
指能表示的**最大與最小數值大小**，由指數位元決定。`float` 的最大值約 1.8×10³⁰⁸，最小正規化正數約 2.2×10⁻³⁰⁸，超過上限會變成 `inf`。

**5. 精度與範圍有何不同？**
- **範圍**＝「尺有多長」（能表示多大、多小的數；由**指數**位元決定）。
- **精度**＝「刻度有多細」（相鄰兩個可表示數之間的相對間距；由**尾數**位元決定）。

兩者是獨立的：範圍很大不代表很精確（如上面 `1e16 + 1 == 1e16`）。浮點數的相鄰間距會隨數值變大而變大，所以越大的數，絕對誤差越大。

## 第六部分：AI 模型中的數值表示

人工智慧模型需要儲存大量參數（Parameters）與權重（Weights）。

請查閱資料並回答：

1. AI 模型中的權重通常使用什麼型別儲存？
2. 什麼是 FP64？
3. 什麼是 FP32？
4. 什麼是 FP16？
5. 什麼是 BF16？
6. 為什麼大型 AI 模型不直接使用 Python 的 `int` 作為權重？

### ✍️ 第六部分：我的作答

先建立實驗工具：以 `struct` 直接檢視 FP64／FP32／FP16 的位元排列，並用「截斷 FP32 的前 16 位元」模擬 BF16（BF16 的定義就是 FP32 的高 16 位元，通常搭配四捨五入）。

In [15]:
import struct
import numpy as np

def to_bf16(x: float) -> float:
    """把數值捨入成 BF16（以 FP32 的高 16 位元表示，round-to-nearest-even）。"""
    b = struct.unpack(">I", struct.pack(">f", x))[0]
    b = (b + 0x7FFF + ((b >> 16) & 1)) & 0xFFFF0000
    return struct.unpack(">f", struct.pack(">I", b))[0]

def layout(x: float, fmt: str):
    """回傳 (符號, 指數, 尾數) 的位元字串。"""
    if fmt == "FP64":
        s = f"{struct.unpack('>Q', struct.pack('>d', x))[0]:064b}"; return s[0], s[1:12], s[12:]
    if fmt == "FP32":
        s = f"{struct.unpack('>I', struct.pack('>f', x))[0]:032b}"; return s[0], s[1:9], s[9:]
    if fmt == "FP16":
        s = f"{struct.unpack('>H', struct.pack('>e', x))[0]:016b}"; return s[0], s[1:6], s[6:]
    if fmt == "BF16":
        b = struct.unpack('>I', struct.pack('>f', to_bf16(x)))[0] >> 16
        s = f"{b:016b}"; return s[0], s[1:9], s[9:]

print("以 0.1 為例的位元排列（符號 | 指數 | 尾數）\n")
for fmt in ["FP64", "FP32", "FP16", "BF16"]:
    sg, ex, ma = layout(0.1, fmt)
    print(f"{fmt}: {sg} | {ex} | {ma}    (1+{len(ex)}+{len(ma)} = {1+len(ex)+len(ma)} bits)")

以 0.1 為例的位元排列（符號 | 指數 | 尾數）

FP64: 0 | 01111111011 | 1001100110011001100110011001100110011001100110011010    (1+11+52 = 64 bits)
FP32: 0 | 01111011 | 10011001100110011001101    (1+8+23 = 32 bits)
FP16: 0 | 01011 | 1001100110    (1+5+10 = 16 bits)
BF16: 0 | 01111011 | 1001101    (1+8+7 = 16 bits)


In [16]:
# 各格式的數值特性
rows = []
for name, dt in [("FP64", np.float64), ("FP32", np.float32), ("FP16", np.float16)]:
    fi = np.finfo(dt)
    rows.append((name, fi.bits, float(fi.max), float(fi.tiny), float(fi.eps)))
bf16_max = struct.unpack(">f", struct.pack(">I", 0x7F7F0000))[0]
bf16_tiny = struct.unpack(">f", struct.pack(">I", 0x00800000))[0]
rows.append(("BF16", 16, bf16_max, bf16_tiny, 2.0**-7))

print(f"{'格式':<6}{'位元':>5}{'最大值':>14}{'最小正規化數':>16}{'epsilon':>14}")
for n, b, mx, tn, ep in rows:
    print(f"{n:<6}{b:>5}{mx:>14.4e}{tn:>16.4e}{ep:>14.4e}")

print("\n同一個數在不同格式下的儲存值：")
for label, val in [("0.1", 0.1), ("π", math.pi)]:
    print(f"\n{label} 的真值    : {val!r}")
    print("  FP64 :", float(np.float64(val)))
    print("  FP32 :", float(np.float32(val)))
    print("  FP16 :", float(np.float16(val)))
    print("  BF16 :", to_bf16(val))

格式       位元           最大值          最小正規化數       epsilon
FP64     64   1.7977e+308     2.2251e-308    2.2204e-16
FP32     32    3.4028e+38      1.1755e-38    1.1921e-07
FP16     16    6.5504e+04      6.1035e-05    9.7656e-04
BF16     16    3.3895e+38      1.1755e-38    7.8125e-03

同一個數在不同格式下的儲存值：

0.1 的真值    : 0.1
  FP64 : 0.1
  FP32 : 0.10000000149011612
  FP16 : 0.0999755859375
  BF16 : 0.10009765625

π 的真值    : 3.141592653589793
  FP64 : 3.141592653589793
  FP32 : 3.1415927410125732
  FP16 : 3.140625
  BF16 : 3.140625


In [17]:
# 範圍差異：FP16 容易溢位，BF16 與 FP32 的範圍相同
with np.errstate(over="ignore", under="ignore"):
    print("100000 → FP16 :", np.float16(100000.0), "（溢位成 inf，FP16 最大 65504）")
    print("100000 → BF16 :", to_bf16(100000.0), "（範圍足夠，但精度較粗）")
    print("1e-8   → FP16 :", np.float16(1e-8),   "（下溢成 0）")
    print("1e-8   → BF16 :", to_bf16(1e-8))

# 精度差異：FP16 的累加會「卡住」
acc16 = np.float16(0); acc32 = np.float32(0)
for _ in range(5000):
    acc16 = np.float16(acc16 + np.float16(1))
    acc32 = np.float32(acc32 + np.float32(1))
print("\n連續加 5000 個 1：FP16 =", acc16, "，FP32 =", acc32)
print("→ 這就是為什麼「混合精度訓練」要用 FP32 做累加與保存主權重。")

100000 → FP16 : inf （溢位成 inf，FP16 最大 65504）
100000 → BF16 : 99840.0 （範圍足夠，但精度較粗）
1e-8   → FP16 : 0.0 （下溢成 0）
1e-8   → BF16 : 1.0011717677116394e-08

連續加 5000 個 1：FP16 = 2.048e+03 ，FP32 = 5000.0
→ 這就是為什麼「混合精度訓練」要用 FP32 做累加與保存主權重。


In [18]:
# 量化：AI 也用「整數」，但是固定位寬的 INT8 + 縮放因子，而不是 Python 的 int
rng = np.random.default_rng(0)
w = rng.normal(0, 0.02, size=1000).astype(np.float32)     # 假想的一層權重

scale = np.abs(w).max() / 127
q = np.round(w / scale).astype(np.int8)                   # INT8 權重
w_hat = q.astype(np.float32) * scale                      # 還原

print("原始權重前 5 個 :", w[:5])
print("INT8 權重前 5 個:", q[:5])
print("還原後前 5 個   :", w_hat[:5])
print("最大量化誤差    :", np.abs(w - w_hat).max())
print("記憶體：FP32 =", w.nbytes, "bytes → INT8 =", q.nbytes, "bytes")

# 為什麼不用 Python int / list 當權重？
N = 1_000_000
py_list = [float(i) for i in range(N)]
np_arr = np.arange(N, dtype=np.float32)
print("\n100 萬個數字的記憶體：")
print("  Python list of float :", (sys.getsizeof(py_list) + sum(sys.getsizeof(v) for v in py_list[:1000]) // 1000 * N) // 1_000_000, "MB（指標 8B + 每個物件 24B）")
print("  NumPy float32 陣列   :", np_arr.nbytes // 1_000_000, "MB（每個數字 4B，連續存放）")

原始權重前 5 個 : [ 0.0025146  -0.0026421   0.01280845  0.002098   -0.01071339]
INT8 權重前 5 個: [  4  -4  21   3 -17]
還原後前 5 個   : [ 0.00245633 -0.00245633  0.01289573  0.00184225 -0.0104394 ]
最大量化誤差    : 0.00030649453
記憶體：FP32 = 4000 bytes → INT8 = 1000 bytes

100 萬個數字的記憶體：
  Python list of float : 32 MB（指標 8B + 每個物件 24B）
  NumPy float32 陣列   : 4 MB（每個數字 4B，連續存放）


**1. AI 模型中的權重通常使用什麼型別儲存？**
使用**浮點數**張量（tensor），而不是 Python 內建的 `int`／`float` 物件。常見有：
- 訓練：**FP32**（傳統預設）、**BF16／FP16 + FP32 主權重**（混合精度訓練，現在的主流做法）。
- 推論：**FP16／BF16**，進一步壓縮則用 **INT8、INT4、FP8** 等量化格式。

**2. 什麼是 FP64？**
雙精度浮點數（double precision，IEEE 754 binary64），共 64 位元：1 位元符號 ＋ 11 位元指數 ＋ 52 位元尾數；約 15～17 位十進位有效數字，範圍約 ±1.8×10³⁰⁸。就是 Python `float` 使用的格式。常用於科學計算、需要高精度之處。

**3. 什麼是 FP32？**
單精度浮點數（binary32），32 位元：1 ＋ 8 ＋ 23；約 6～9 位十進位有效數字，範圍約 ±3.4×10³⁸。是傳統深度學習訓練的預設精度。

**4. 什麼是 FP16？**
半精度浮點數（binary16），16 位元：1 ＋ 5 ＋ 10；約 3～4 位有效數字，**最大值只有 65504**，容易溢位／下溢，訓練時常需搭配 loss scaling。記憶體與頻寬只有 FP32 的一半。

**5. 什麼是 BF16？**
Brain Floating Point（Google Brain 提出），16 位元：1 ＋ **8** ＋ **7**。指數位元與 FP32 相同（所以**範圍與 FP32 一樣大**），但尾數只有 7 位（精度比 FP16 更粗）。它等於「把 FP32 的後 16 位元砍掉」，與 FP32 互轉很容易，訓練時比 FP16 穩定（不易溢位），已成為大型模型訓練的主流。

**6. 為什麼大型 AI 模型不直接使用 Python 的 `int` 作為權重？**
1. **權重是小數**：權重通常是很小的實數（如 0.013、-0.27），`int` 無法表示小數部分（除非另加縮放因子做量化）。
2. **記憶體太浪費**：每個 Python `int` 是一個完整的物件（至少約 28 bytes ＋ 串列中的 8-byte 指標），而 FP16 只要 2 bytes；千億個參數根本放不下。
3. **速度太慢**：Python 物件分散在記憶體、要一個個解讀型別，無法使用 GPU／TPU 的向量化與矩陣乘法硬體（Tensor Core）。
4. **任意精度沒有好處**：神經網路對微小誤差不敏感，沒有必要花大量資源換取「精確」。
5. 補充：AI 確實使用**整數**（INT8、INT4），但那是**固定位寬**、配合縮放因子的量化格式，與 Python 任意精度 `int` 完全不同。

## 第七部分：比較不同數值表示方式

完成下表。

| 表示方式 | 位元數 | 是否有誤差 | 主要用途 |
|----------|--------|------------|----------|
| int | ? | ? | ? |
| FP64 | ? | ? | ? |
| FP32 | ? | ? | ? |
| FP16 | ? | ? | ? |
| BF16 | ? | ? | ? |

並回答：

1. 哪種表示方式最精確？
2. 哪種表示方式最節省記憶體？
3. 哪種表示方式最常見於 AI 訓練？
4. 哪種表示方式最常見於 AI 推論（Inference）？

### ✍️ 第七部分：我的作答

| 表示方式 | 位元數 | 是否有誤差 | 主要用途 |
|----------|--------|------------|----------|
| int | Python `int`：**不固定**（任意精度）；硬體／NumPy／AI 的整數為固定位寬：INT8＝8、INT32＝32、INT64＝64 | Python `int` 的整數運算**沒有誤差**（精確）；固定位寬整數有溢位風險；把實數「量化」成 INT8 則有量化誤差 | 計數、索引、token id、迴圈、密碼學／大數運算；INT8／INT4 用於量化推論 |
| FP64 | **64**（1＋11＋52） | 有（約 15～17 位有效數字，機器 ε ≈ 2.2×10⁻¹⁶） | 科學計算、氣象與物理模擬、需要高精度的數值分析 |
| FP32 | **32**（1＋8＋23） | 有（約 6～9 位有效數字，ε ≈ 1.2×10⁻⁷） | 傳統深度學習訓練的預設精度；混合精度中的主權重與累加 |
| FP16 | **16**（1＋5＋10） | 有（約 3～4 位有效數字，ε ≈ 9.8×10⁻⁴），範圍小（最大 65504） | 混合精度訓練、模型推論、圖形處理 |
| BF16 | **16**（1＋8＋7） | 有，最粗（約 2～3 位有效數字，ε ≈ 7.8×10⁻³），但範圍與 FP32 相同 | 大型語言模型訓練與推論（TPU、新世代 GPU 都支援） |

**問題回答**

1. **哪種表示方式最精確？**
   在浮點格式中是 **FP64**。若只看整數，Python 的任意精度 **`int`** 最精確（無誤差）。
2. **哪種表示方式最節省記憶體？**
   在表中的浮點格式裡是 **FP16 與 BF16**（各 16 位元＝2 bytes，是 FP32 的一半、FP64 的四分之一）；若把 INT8／INT4 量化格式也算進來，則 INT8（1 byte）、INT4（0.5 byte）更省。
3. **哪種表示方式最常見於 AI 訓練？**
   以 **混合精度訓練** 為主：用 **BF16（或 FP16）** 做前向／反向計算，用 **FP32** 保存主權重與累加。目前大型語言模型訓練以 **BF16** 最常見；FP32 是傳統預設。
4. **哪種表示方式最常見於 AI 推論（Inference）？**
   **FP16／BF16**，再進一步常用 **INT8、INT4（量化）** 甚至 FP8，以降低記憶體與延遲。

## 第八部分：綜合思考

### 思考題一

Python 官方文件指出：

```python
int
```

具有 Unlimited Precision（無限制精度）。

請回答：

1. Unlimited Precision 是什麼意思？
2. 它是否代表可以儲存無限大的整數？
3. CPython 如何實現這項特性？

### ✍️ 第八部分：思考題一 —— 我的作答

In [19]:
import math

f = math.factorial(1000)
print("1000! 的位數            :", len(str(f)))
print("1000! 的 bit_length     :", f.bit_length())
print("1000! 的物件大小        :", sys.getsizeof(f), "bytes")

g = 2 ** 1_000_000
print("\n2**1,000,000 的 bit_length:", g.bit_length())
print("2**1,000,000 的物件大小   :", sys.getsizeof(g) // 1024, "KB")
print("每 30 bits 約 4 bytes，所以大小 ≈ bit_length / 30 × 4 =", g.bit_length() // 30 * 4 // 1024, "KB")

# 每個 digit 的基底是 2**30：用 divmod 把整數拆成 digit
n = 2**65 + 12345
digits, m = [], n
while m:
    m, d = divmod(m, 2**30)
    digits.append(d)
print("\n2**65 + 12345 拆成 base-2**30 的 digit（低位在前）:", digits)

1000! 的位數            : 2568
1000! 的 bit_length     : 8530
1000! 的物件大小        : 1164 bytes

2**1,000,000 的 bit_length: 1000001
2**1,000,000 的物件大小   : 130 KB
每 30 bits 約 4 bytes，所以大小 ≈ bit_length / 30 × 4 = 130 KB

2**65 + 12345 拆成 base-2**30 的 digit（低位在前）: [12345, 0, 32]


**1. Unlimited Precision 是什麼意思？**
整數運算的結果**永遠精確**，Python 不會因為位數增加而截斷、捨入或溢位。程式設計師不必像 C 語言那樣選擇 `int`／`long`／`long long`，也不用擔心 `int32` 溢位。

**2. 它是否代表可以儲存無限大的整數？** —— **不是**。「Unlimited」是指語言**沒有預設的位元數上限**，但實際上仍受限於：
- 電腦的**記憶體**（整數越大占的 digit 越多）；
- **計算時間**（位數越多越慢）；
- 實作內部欄位（如 digit 數量）所能表示的極限。

宇宙中的原子數都比「無限大」小得多，所以電腦當然不可能真正儲存無限大的整數。

**3. CPython 如何實現這項特性？**
- 以 `PyLongObject` 表示整數：**符號＋長度＋變長 digit 陣列**，每個 digit 是 **30 bits**（以 2³⁰ 為基底，存在 32-bit 欄位中，留出 2 個位元讓進位運算不會溢位）。
- 運算時，加減乘除依「小學直式運算」逐個 digit 處理進位／借位；位數很大時乘法改用 **Karatsuba** 演算法加速。
- 整數是**不可變**的，運算後產生新的物件，長度依結果需要配置；因此不會溢位，只會占用更多記憶體。

### 思考題二

比較：

```python
int
```

與

```python
float
```

回答：

1. 為什麼 `int` 幾乎沒有固定上限？
2. 為什麼 `float` 卻存在範圍限制？
3. 為什麼 `float` 可能產生誤差？

### ✍️ 第八部分：思考題二 —— 我的作答

In [20]:
# int：位數自動增加；float：固定 64 bits
print("int  : 2**100     =", 2**100)
print("float: 2.0**100   =", 2.0**100, "(只保留約 16 位有效數字)")
print("兩者差異 :", 2**100 - int(2.0**100), "（2 的次方剛好能精確表示，改用 10**30 看看 ↓）")
print("int(1e30) =", int(1e30), "  vs  10**30 =", 10**30, "→ 相等嗎？", int(1e30) == 10**30)

print("\nfloat 的範圍限制：")
print("  float 最大值 :", sys.float_info.max)
print("  1e308 * 10   :", 1e308 * 10)
print("\nfloat 的誤差：")
print("  0.1 * 3      :", 0.1 * 3, "→ ==0.3 ?", 0.1 * 3 == 0.3)
print("  2**53 + 1 (float) :", float(2**53) + 1, "→ 加 1 沒有效果")
print("  2**53 + 1 (int)   :", 2**53 + 1)

int  : 2**100     = 1267650600228229401496703205376
float: 2.0**100   = 1.2676506002282294e+30 (只保留約 16 位有效數字)
兩者差異 : 0 （2 的次方剛好能精確表示，改用 10**30 看看 ↓）
int(1e30) = 1000000000000000019884624838656   vs  10**30 = 1000000000000000000000000000000 → 相等嗎？ False

float 的範圍限制：
  float 最大值 : 1.7976931348623157e+308
  1e308 * 10   : inf

float 的誤差：
  0.1 * 3      : 0.30000000000000004 → ==0.3 ? False
  2**53 + 1 (float) : 9007199254740992.0 → 加 1 沒有效果
  2**53 + 1 (int)   : 9007199254740993


**1. 為什麼 `int` 幾乎沒有固定上限？**
因為 `int` 以**可變長度的 digit 陣列**儲存，需要多大就配置多大；每次運算都會依結果重新決定長度，所以不會像固定位寬的整數一樣溢位。

**2. 為什麼 `float` 卻存在範圍限制？**
`float` 是**固定 64 位元**（1＋11＋52）的格式，指數只有 11 位元，能表示的指數大約 -1022～+1023，因此最大約 1.8×10³⁰⁸；超過就變成 `inf` 或丟出 `OverflowError`（例如 `float(10**400)`）。這是為了讓 CPU 的浮點運算單元（FPU）能在**一個指令週期內完成**運算——用「固定大小」換取「速度」。

**3. 為什麼 `float` 可能產生誤差？**
因為 52 位元的尾數只能保留約 15～17 位十進位有效數字，而且是**二進位**表示：許多十進位小數（如 0.1）在二進位是無限循環，只能存最接近的近似值；運算結果也會再被四捨五入回 53 位元有效數字，所以每次運算都可能引入微小誤差，且誤差會累積。

### 思考題三

比較下列兩種設計理念：

#### Python 的整數

```text
Arbitrary Precision
正確性優先
```

#### AI 模型常用的 FP16、BF16

```text
Limited Precision
效率優先
```

請說明：

1. 兩種設計理念的差異。
2. 各自的優點與缺點。
3. 為什麼大型語言模型通常選擇較低精度的表示方式？
4. 如果所有 AI 模型都改用 FP64，可能會產生哪些問題？

### ✍️ 第八部分：思考題三 —— 我的作答

| 面向 | Python 整數（Arbitrary Precision） | FP16／BF16（Limited Precision） |
|------|-----------------------------------|--------------------------------|
| 設計目標 | **正確性優先**：結果永遠精確 | **效率優先**：用最少的位元、最快的速度得到「夠用」的答案 |
| 大小 | 不固定，隨數值增長 | 固定 16 位元（2 bytes） |
| 誤差 | 無誤差 | 有捨入誤差（BF16 約 3 位十進位有效數字） |
| 速度／硬體 | 以軟體逐 digit 計算，無法直接用 GPU 向量硬體 | 由 GPU／TPU 硬體直接支援，可大量平行 |
| 記憶體 | 大整數占用多，且每個都是物件 | 每個數只占 2 bytes，緊密連續排列 |
| 適合場景 | 密碼學（RSA）、精確計數、數論、大數運算 | 神經網路訓練與推論、圖形處理 |

**1. 兩種設計理念的差異：**
前者要求「每一個位數都對」，後者接受「大致正確就好」，換取速度與空間。取捨的依據是**問題本身對誤差的容忍度**：密碼學一個位元錯就全錯；神經網路有數十億個參數，且訓練本身就帶有隨機性（隨機初始化、SGD 雜訊），輕微的捨入誤差幾乎不影響最後的表現。

**2. 各自的優點與缺點：**
- 任意精度整數：✅ 精確、不溢位、使用簡單；❌ 慢、耗記憶體、不能表示小數、無法用 GPU 加速。
- FP16／BF16：✅ 省記憶體與頻寬、GPU／TPU 極快、能大幅擴大可訓練的模型大小；❌ 精度低、可能溢位／下溢（FP16）、累加誤差大，需搭配 FP32 累加或 loss scaling 等技巧。

**3. 為什麼大型語言模型通常選擇較低精度？**
- **模型太大**：數十億到數千億個參數，記憶體是第一瓶頸（見挑戰題的計算）。
- **記憶體頻寬瓶頸**：GPU 算得比資料搬運還快，資料越小、搬得越快，整體越快。
- **神經網路本來就對雜訊有容忍度**：少幾位有效數字對準確率影響很小。
- **硬體支援**：GPU 的 Tensor Core、TPU 針對 FP16／BF16／INT8 有專門的矩陣乘法單元，吞吐量遠高於 FP64。
- **BF16 擁有與 FP32 相同的範圍**，不易溢位，訓練穩定。

**4. 如果所有 AI 模型都改用 FP64，可能會產生哪些問題？**
- 記憶體需求是 FP32 的 2 倍、FP16／BF16 的 4 倍：原本一張 GPU 能放下的模型可能要用 4 張。
- 計算速度大幅下降：很多 GPU 的 FP64 吞吐量遠低於低精度，且無法使用 Tensor Core 的低精度加速。
- 資料傳輸量加倍：GPU 與記憶體之間、GPU 與 GPU 之間的頻寬吃緊，訓練時間與電力成本上升（金錢與碳排放）。
- 同時 KV cache、激活值（activations）、梯度全部變大，推論延遲與成本增加。
- 換來的精度提升對模型品質**幾乎沒有幫助**，成本效益極差。

### 挑戰題

請用自己的話解釋：

```python
x = 10**1000
```

為什麼 Python 可以正常運作；

而大型 AI 模型卻通常選擇使用：

```text
FP32
FP16
BF16
```

而不是任意精度整數或高精度浮點數？

請從：

- 記憶體使用量
- 計算速度
- 數值精度
- GPU 運算效率

等角度進行討論。

### ✍️ 挑戰題 —— 我的作答

先用數字說明「為什麼要在意每個參數多幾個 bit」。

In [21]:
import time

# (1) 記憶體使用量：只是「存放權重」需要多少 GB（不含激活值、梯度、優化器狀態）
formats = [("FP64", 8), ("FP32", 4), ("FP16", 2), ("BF16", 2), ("INT8", 1), ("INT4", 0.5)]
print(f"{'格式':<6}{'bytes/參數':>11}{'7B 模型':>12}{'70B 模型':>12}{'405B 模型':>13}")
for name, b in formats:
    print(f"{name:<6}{b:>11}{7e9*b/1e9:>10.1f}GB{70e9*b/1e9:>10.1f}GB{405e9*b/1e9:>11.1f}GB")
print("（GPU 記憶體常見為 24GB～80GB，單卡放不下 70B 的 FP32 模型）")

# (2) 計算速度：Python 逐個處理物件 vs. 向量化的連續陣列（結果依機器而異）
N = 2_000_000
a_list = [float(i) for i in range(N)]
b_list = [float(i) for i in range(N)]
a_np = np.arange(N, dtype=np.float32); b_np = np.arange(N, dtype=np.float32)

t0 = time.perf_counter(); r1 = [x * y for x, y in zip(a_list, b_list)]; t_py = time.perf_counter() - t0
t0 = time.perf_counter(); r2 = a_np * b_np;                             t_np = time.perf_counter() - t0
print(f"\n{N:,} 次乘法：Python list = {t_py*1000:.1f} ms，NumPy float32 = {t_np*1000:.1f} ms，約快 {t_py/t_np:.0f} 倍")

# (3) 數值精度：低精度的誤差有多大？
x = 1 / 3
print("\n1/3 在各格式下的誤差：")
for name, val in [("FP64", float(np.float64(x))), ("FP32", float(np.float32(x))),
                  ("FP16", float(np.float16(x))), ("BF16", to_bf16(x))]:
    print(f"  {name}: {val:.12f}   相對誤差 = {abs(val - x) / x:.2e}")

# (4) 任意精度整數 x = 10**1000 的代價
x = 10**1000
print(f"\n10**1000 需要 {sys.getsizeof(x)} bytes 才能『精確』存一個數；FP16 只要 2 bytes（但只有約 3 位有效數字，且最大 65504）")

格式       bytes/參數       7B 模型      70B 模型      405B 模型
FP64            8      56.0GB     560.0GB     3240.0GB
FP32            4      28.0GB     280.0GB     1620.0GB
FP16            2      14.0GB     140.0GB      810.0GB
BF16            2      14.0GB     140.0GB      810.0GB
INT8            1       7.0GB      70.0GB      405.0GB
INT4          0.5       3.5GB      35.0GB      202.5GB
（GPU 記憶體常見為 24GB～80GB，單卡放不下 70B 的 FP32 模型）

2,000,000 次乘法：Python list = 79.8 ms，NumPy float32 = 2.6 ms，約快 31 倍

1/3 在各格式下的誤差：
  FP64: 0.333333333333   相對誤差 = 0.00e+00
  FP32: 0.333333343267   相對誤差 = 2.98e-08
  FP16: 0.333251953125   相對誤差 = 2.44e-04
  BF16: 0.333984375000   相對誤差 = 1.95e-03

10**1000 需要 468 bytes 才能『精確』存一個數；FP16 只要 2 bytes（但只有約 3 位有效數字，且最大 65504）


**`x = 10**1000` 為什麼在 Python 可以正常運作？**
因為 Python 的 `int` 是**任意精度**：CPython 用 `PyLongObject` 依需要配置一串 30-bit 的 digit（`10**1000` 約 111 個 digit、約 460 bytes），並以軟體逐位計算，結果**永遠精確**。這是一個「正確性優先」的設計：程式很少同時處理數十億個這樣的數，所以每個數多花幾百 bytes、慢一點都可以接受。

**為什麼大型 AI 模型卻通常選擇 FP32／FP16／BF16，而不是任意精度整數或高精度浮點數？**
因為 AI 的處境和上面完全相反：**要同時處理數十億到數千億個數字，而且要重複計算無數次**。此時「每個數字花多少 bytes、多少時間」會被放大數十億倍。從四個角度看：

1. **記憶體使用量**
   一個 70B 參數的模型，用 FP64 需約 560 GB、FP32 約 280 GB、FP16／BF16 約 140 GB、INT4 僅約 35 GB（見上方計算）。任意精度物件連 Python 的物件標頭都比 FP16 大十幾倍。降低精度讓模型能放進較少的 GPU，是「放得下 vs. 放不下」的差別；訓練時還要加上梯度與優化器狀態（Adam 需要額外好幾倍），壓力更大。

2. **計算速度**
   神經網路的核心運算是巨量矩陣乘法。GPU／TPU 的矩陣運算硬體（如 Tensor Core）為 FP16／BF16／INT8 等低精度特別設計，吞吐量遠高於 FP64。此外，GPU 常常是**被記憶體頻寬卡住**而不是被算力卡住：資料越小、搬得越快，整體越快。任意精度整數則必須由軟體逐個 digit 運算，根本無法平行化到這種程度（上面 Python vs. NumPy 的實驗就已差了數十倍）。

3. **數值精度**
   神經網路對微小誤差有很高的容忍度：權重只是「大致正確的強度」，訓練本身帶有隨機雜訊，少幾位有效數字不會明顯影響準確率。既然精度「多出來的部分」對結果沒有幫助，多付出的記憶體與時間就是浪費。但精度也不能太低：所以採用**混合精度**——用 BF16／FP16 做大量計算，用 FP32 做累加與保存主權重（上面示範了 FP16 累加會卡在 2048）；BF16 保留 FP32 的範圍，避免溢位。

4. **GPU 運算效率**
   GPU 靠**大量規則、同型、連續存放**的資料做平行運算（SIMD／SIMT）。固定位寬的 FP16／BF16 每個數字大小一致、可整齊排列、一次搬移／計算大量元素；Python 的 `int` 是變長物件，每個數字長度不同、要逐個解讀，無法發揮 GPU 的優勢。同一片晶片面積上，低精度的乘加單元也比高精度小得多，因此能塞進更多運算單元，總吞吐量更高。

**結論：**
Python 的 `int` 追求「**每一個 bit 都要正確**」，適合少量但要求精確的計算（密碼學、數論、精確計數）；AI 模型追求「**用最少資源得到夠好的答案**」，適合海量但容忍誤差的計算。兩者沒有誰比較好，而是**依問題需求做出不同的取捨**。

> 一句話總結：**AI 選擇「夠用就好的低精度」，不是因為低精度比較準，而是因為在「模型大小、訓練速度、硬體效率」三者的總成本下，它是最划算的選擇。**